### DISCUSSION TITLE
Light GCN implementation

### DISCUSSION THREAD

**Discussion by nishithshowri006** [pyg-team/pytorch_geometric#3992] [Q&A]
Hey I am newbie to geometric trying to implement Light GCN but it's so confusing and I am not able to do so. Also I am unable to understand how to use the Movie Lens dataset in Datasets its not at all clear. Is there any tutorial or pipeline I could check so that I can work on.

**Comment by rusty1s** [comment]
Hey, thanks for your interest. We are still working on a LightGCN example. You can find a preliminary version [here](https://colab.research.google.com/drive/1o0EuldXGwcz0j66n8sKT2yA5UQzkkddA?usp=sharing). Hope this is useful to you!

**Reply by nishithshowri006** [reply]
Hey thanks for the quick response, with some help from this I trained LightGCN on movie lens data set but it is performing significantly underperforming than the LightGCNs of vinila pytorch on the same movie lens data.

**Reply by rusty1s** [reply]
Is this on the exact same dataset? Do you have a reference implementation?

**Reply by nishithshowri006** [reply]
Here is the implementation I made [notebook link](https://colab.research.google.com/drive/1bqtnArdMBTYX2rwrP9yr-Dd6RJ9Eiebh?usp=sharing)

**Reply by rusty1s** [reply]
Very interesting. I will try to study this in more depth. Any help/pointers on what might be different in both implementations is highly appreciated :)

**Reply by rusty1s** [reply]
I checked both versions, and besides a different initialization technique (your version uses `xavier_uniform_`) and applying dropout between different rounds of propagations (this is `0` in your case anyway), both versions yield the exact same output embeddings. Note that the difference in speedup can probably be explained by this line:
```python
        rankings_pos = torch.tensor([
            torch.sum(embed[users[i]] * embed[pos_items[i]])
            for i in range(lens)
        ])
        rankings_neg = torch.tensor([
            torch.sum(embed[users[i]] * embed[neg_items[i]])
            for i in range(lens)
        ])
```
which uses Python loops to gather all edge-level embeddings.

**Reply by rusty1s** [reply]
They are only the same in case you set `torch.manual_seed(12345)` right before the `torch.nn.Embedding` construction (and remove the `reset_parameters()` call in `LightGCN`). I definitely agree that there exists a problem but it needs to be caused by something different than the model implementation.

**Reply by nishithshowri006** [reply]
> They are only the same in case you set `torch.manual_seed(12345)` right before the `torch.nn.Embedding` construction (and remove the `reset_parameters()` call in `LightGCN`). I definitely agree that there exists a problem but it needs to be caused by something different than the model implementation.

Hey so yeah when I did add the Xavier initialisation to the pyg model I got almost similar results..

**Reply by rusty1s** [reply]
Super interesting. Shall we fix the initialization on the PyG `LightGCN` side? Are you interested in contributing this?

**Reply by nishithshowri006** [reply]
Yeah sure we can do that..yeah sure I can contribute to this

**Reply by rusty1s** [reply]
Thanks for your contributions. Fixed via https://github.com/pyg-team/pytorch_geometric/pull/4083.

**Reply by nishithshowri006** [reply]
You are welcome

**Comment by nishithshowri006** [comment]
Hey @rusty1s is there any way to contribute my notebook as an example?

**Reply by rusty1s** [reply]
Definitely, but I prefer short one-file examples scripts rather than notebooks (makes version control and reviewing easier). For now, we could start by just adding all notebook cells into a single Python script. WDYT?

**Reply by rusty1s** [reply]
Added an issue to track progress: https://github.com/pyg-team/pytorch_geometric/issues/4182
